File: 27-Trie/python/02_trie_implementation_and_advanced_operations.py
class Node:
    def __init__(self):
        self.children = [None] * 26
        self.end_count = 0
        self.prefix_count = 0

class Trie:
    def __init__(self):
        self.root = Node()

    def insert(self, word: str) -> None:
        curr = self.root
        for c in word:
            idx = ord(c) - ord('a')
            if not curr.children[idx]:
                curr.children[idx] = Node()
            curr = curr.children[idx]
            curr.prefix_count += 1
        curr.end_count += 1

    def countWordsEqualTo(self, word: str) -> int:
        curr = self.root
        for c in word:
            idx = ord(c) - ord('a')
            if not curr.children[idx]:
                return 0
            curr = curr.children[idx]
        return curr.end_count

    def countWordsStartingWith(self, prefix: str) -> int:
        curr = self.root
        for c in prefix:
            idx = ord(c) - ord('a')
            if not curr.children[idx]:
                return 0
            curr = curr.children[idx]
        return curr.prefix_count

    def erase(self, word: str) -> None:
        if self.countWordsEqualTo(word) == 0:
            return
        curr = self.root
        for c in word:
            idx = ord(c) - ord('a')
            curr = curr.children[idx]
            curr.prefix_count -= 1
        curr.end_count -= 1

File: 27-Trie/java/02-TrieImplementationAndAdvancedOperations.java
class Trie {
    static class Node {
        Node[] children = new Node[26];
        int endCount = 0;
        int prefixCount = 0;
    }

    private Node root;

    public Trie() {
        root = new Node();
    }

    public void insert(String word) {
        Node curr = root;
        for (char c : word.toCharArray()) {
            int idx = c - 'a';
            if (curr.children[idx] == null) curr.children[idx] = new Node();
            curr = curr.children[idx];
            curr.prefixCount++;
        }
        curr.endCount++;
    }

    public int countWordsEqualTo(String word) {
        Node curr = root;
        for (char c : word.toCharArray()) {
            int idx = c - 'a';
            if (curr.children[idx] == null) return 0;
            curr = curr.children[idx];
        }
        return curr.endCount;
    }

    public int countWordsStartingWith(String prefix) {
        Node curr = root;
        for (char c : prefix.toCharArray()) {
            int idx = c - 'a';
            if (curr.children[idx] == null) return 0;
            curr = curr.children[idx];
        }
        return curr.prefixCount;
    }

    public void erase(String word) {
        if (countWordsEqualTo(word) == 0) return;
        Node curr = root;
        for (char c : word.toCharArray()) {
            int idx = c - 'a';
            curr = curr.children[idx];
            curr.prefixCount--;
        }
        curr.endCount--;
    }
}
